src/WebJobs.Script/Diagnostics/FileLogger.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Diagnostics
+{
+    /// <summary>
+    /// A wrapper class to allow ILoggers from functions to write to the existing FileTraceWriter. This allows
+    /// logs to show up in the log files and stream in the portal.
+    /// </summary>
+    internal class FileLogger : ILogger
+    {
+        private static readonly ConcurrentDictionary<string, TraceWriter> _writerCache = new ConcurrentDictionary<string, TraceWriter>();
+        private readonly Func<string, LogLevel, bool> _filter;
+        private readonly string _categoryName;
+        private readonly ScriptHostConfiguration _config;
+
+        public FileLogger(string categoryName, ScriptHostConfiguration config, Func<string, LogLevel, bool> filter)
+        {
+            _categoryName = categoryName;
+            _config = config;
+            _filter = filter;
+        }
+
+        public IDisposable BeginScope<TState>(TState state) => DictionaryLoggerScope.Push(state);
+
+        public bool IsEnabled(LogLevel logLevel) => _filter(_categoryName, logLevel);
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        {
+            // We don't support any other type of state.
+            IEnumerable<KeyValuePair<string, object>> properties = state as IEnumerable<KeyValuePair<string, object>>;
+            string formattedMessage = formatter?.Invoke(state, exception);
+
+            if (string.IsNullOrEmpty(formattedMessage) || properties == null || !IsEnabled(logLevel) || IsFromTraceWriter(properties))
+            {
+                return;
+            }
+
+            IDictionary<string, object> scopeProperties = DictionaryLoggerScope.GetMergedStateDictionary();
+            string functionName = scopeProperties["MS_FunctionName"].ToString();
+
+            // this function name starts with "Function.", but file paths do not include this
+            string functionsPrefix = "Functions.";
+            if (functionName.StartsWith(functionsPrefix))
+            {
+                functionName = functionName.Substring(functionsPrefix.Length);
+            }
+
+            TraceWriter traceWriter = _writerCache.GetOrAdd(functionName, (n) => CreateFileTraceWriter(n, _config));
+            TraceLevel traceLevel = GetTraceLevel(logLevel);
+            TraceEvent traceEvent = new TraceEvent(traceLevel, formattedMessage, _categoryName, exception);
+
+            // FileTraceWriter doesn't use these properties, but adding them for completeness
+
+            traceWriter.Trace(traceEvent);
```

Comment:
I filter above -- if this came from the user's TraceWriter, it doesn't make it through to here. If it's from the ILogger, it does. User logging is the one case where using the traceWriter will log to both ILogger and TraceWriter.

I mentioned it above -- I wanted to avoid this as well by have a clean separation between ILogger and TraceWriter. But from a user's perspective, now they will automatically get App Insights with their existing traces... which is nice, but leads to some complexity when doing things like this.